service/udp.go

```diff
@@ -321,92 +320,239 @@ func (c *natconn) ReadFrom(buf []byte) (int, net.Addr, error) {
-		metrics:        connMetrics,
-		defaultTimeout: m.timeout,
+// Del deletes a UDP NAT entry from the natmap.
+func (m *natmap) Del(clientAddr string) {
+	m.Lock()
+	defer m.Unlock()
+
+	if _, ok := m.associations[clientAddr]; ok {
+		delete(m.associations, clientAddr)
 	}
+}
 
+// Add adds a new UDP NAT entry to the natmap.
+func (m *natmap) Add(clientAddr string, assoc PacketAssociation) {
 	m.Lock()
 	defer m.Unlock()
 
-	m.keyConn[key] = entry
-	return entry
+	m.associations[clientAddr] = assoc
 }
 
-func (m *natmap) del(key string) net.PacketConn {
+func (m *natmap) Close() error {
 	m.Lock()
 	defer m.Unlock()
 
-	entry, ok := m.keyConn[key]
-	if ok {
-		delete(m.keyConn, key)
-		return entry
+	var err error
+	for _, assoc := range m.associations {
+		if e := assoc.Close(); e != nil {
+			err = e
+		}
 	}
-	return nil
+	return err
 }
 
-func (m *natmap) Add(clientAddr net.Addr, clientConn net.PacketConn, cryptoKey *shadowsocks.EncryptionKey, targetConn net.PacketConn, keyID string) *natconn {
-	connMetrics := m.metrics.AddUDPNatEntry(clientAddr, keyID)
-	entry := m.set(clientAddr.String(), targetConn, cryptoKey, connMetrics)
+// ConnAssociation represents a UDP association that handles incoming and
+// outgoing packets from a [net.Conn] and forwards them to a target connection,
+// and vice versa. Used by Caddy.
+type ConnAssociation interface {
+	// Handle reads data from the association and handles incoming and outgoing
+	// packets.
+	Handle()
+}
 
-	go func() {
-		timedCopy(clientAddr, clientConn, entry, m.logger)
-		connMetrics.RemoveNatEntry()
-		if pc := m.del(clientAddr.String()); pc != nil {
-			pc.Close()
+// PacketAssociation represents a UDP association that handles individual
+// incoming packets. Used by outline-ss-server.
+type PacketAssociation interface {
+	// HandlePacket processes a single incoming packet.
+	//
+	// pkt contains the raw packet data.
+	// lazySlice is the LazySlice that holds the pkt buffer, which should be
+	// released after the packet is processed.
+	HandlePacket(pkt []byte, lazySlice slicepool.LazySlice)
+
+	// Done returns a channel that is closed when the association is closed.
+	Done() <-chan struct{}
+
+	// Close closes the association and releases any associated resources.
+	Close() error
+}
+
+type association struct {
+	net.Conn
+	raddr             net.UDPAddr
+	m                 UDPAssociationMetrics
+	logger            *slog.Logger
+	targetConn        net.PacketConn
+	cryptoKey         *shadowsocks.EncryptionKey
+	bufPool           *slicepool.Pool
+	ciphers           CipherList
+	ssm               ShadowsocksConnMetrics
+	targetIPValidator onet.TargetIPValidator
+	doneCh            chan struct{}
+	findAccessKeyOnce sync.Once
+}
+
+var _ ConnAssociation = (*association)(nil)
+var _ PacketAssociation = (*association)(nil)
+
+func (a *association) debugLog(template string, attrs ...slog.Attr) {
+	debugUDP(a.logger, template, attrs...)
+}
+
+func (a *association) Handle() {
```

Comment:
Move this to a HandleAssociation(Conn) association method. I believe the buffer pool can also be global, rather than tied to the association.